x/stakeibc/keeper/reward_converter_test.go

```diff
@@ -0,0 +1,425 @@
+	requestData := query.RequestData[1:] // Remove BalancePrefix byte
+	actualAddress, actualDenom, err := banktypes.AddressAndDenomFromBalancesStore(requestData)
+	s.Require().NoError(err, "no error expected when retrieving address and denom from store key")
+	s.Require().Equal(tradeRoute.TradeAccount.Address, actualAddress.String(), "query account address")
+	s.Require().Equal(tradeRoute.HostDenomOnTradeZone, actualDenom, "query denom")
+}
+
+// Tests a TradeConvertedBalanceQuery that fails due to an invalid account address
+func (s *KeeperTestSuite) TestTradeConvertedBalanceQuery_Failure_InvalidAccountAddress() {
+	tradeRoute := s.SetupTradeConvertedBalanceQueryTestCase()
+
+	// Change the trade ICA account address to be invalid
+	tradeRoute.TradeAccount.Address = "invalid_address"
+
+	err := s.App.StakeibcKeeper.TradeConvertedBalanceQuery(s.Ctx, tradeRoute)
+	s.Require().ErrorContains(err, "invalid trade account address")
+}
+
+
+// Tests a TradeConvertedBalanceQuery that fails due to a missing epoch tracker
+func (s *KeeperTestSuite) TestTradeConvertedBalanceQuery_Failure_MissingEpoch() {
+	tradeRoute := s.SetupTradeConvertedBalanceQueryTestCase()
+
+	// Remove the stride epoch so the test fails
+	s.App.StakeibcKeeper.RemoveEpochTracker(s.Ctx, epochtypes.STRIDE_EPOCH)
+
+	err := s.App.StakeibcKeeper.TradeConvertedBalanceQuery(s.Ctx, tradeRoute)
+	s.Require().ErrorContains(err, "stride_epoch: epoch not found")
+}
+
+// Tests a TradeConvertedBalanceQuery that fails to submit the query due to bad connection
+func (s *KeeperTestSuite) TestTradeConvertedBalanceQuery_FailedQuerySubmission() {
+	tradeRoute := s.SetupTradeConvertedBalanceQueryTestCase()
+
+	// Change the trade ICA connection id to be invalid
+	tradeRoute.TradeAccount.ConnectionId = "invalid_connection"
+
+	err := s.App.StakeibcKeeper.TradeConvertedBalanceQuery(s.Ctx, tradeRoute)
+	s.Require().ErrorContains(err, "invalid connection-id (invalid_connection)")
+}
+
+
+
+
+
+
+
+// The functions which build and send off the ICAs for token transfers and swap commands
+// all use the SubmitTxWithoutCallback, so no callback data is stored to examine
+// We only need to test that an IBC message was indeed sent
+
+// Since each of the ICA functions is called from the icqcallback functions
+// We can test these callback functions directly and implictly test the ICA building funcs
+
+// Useful structures used by the icqcallback tests
+type ICQCallbackArgs struct {
+	Query        icqtypes.Query
+	CallbackArgs []byte
+}
+
+type BalanceQueryCallbackTestCase struct {
+	TradeRoute	types.TradeRoute
+	Response	ICQCallbackArgs
+	ChannelID	string
+	PortID		string
+}
+
+// The callback tests are in their own appropriately named files, similar to other callback tests
+// WithdrawalRewardBalanceCallback will trigger TransferRewardTokensHostToTrade
+// TradeRewardBalanceCallback will trigger SwapRewardTokens
+// TradeConvertedBalanceCallback will trigger TransferConvertedTokensTradeToHost
```

Comment:
were these comments scratch work? also, you can move the types up to the top of the file